### PR TITLE
Feature manga viewer

### PR DESCRIPTION
**Pull request by AS2**

### CODE REVIEW

File: src/MangaJet/app/src/main/res/layout/manga_reader_pager.xml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:text="TextView"
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent"
+        app:layout_constraintVertical_bias="0.92" />
+
+</androidx.constraintlayout.widget.ConstraintLayout>
```
Comment (Sairsey) [inline src/MangaJet/app/src/main/res/layout/manga_reader_pager.xml:34]: Line

File: src/MangaJet/app/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -20,4 +20,9 @@
 
     <string name="aboutMangaFragmentTagsText">Tags</string>
     <string name="aboutMangaDescriptionText">Description</string>
+
+    <!-- Strings, using in "Manga reader" activity -->
+    <string name="manga_reader_bad_page">Ca</string>
```
Comment (Sairsey) [inline src/MangaJet/app/src/main/res/values/strings.xml:25]: String fix

**Review by Sairsey — CHANGES_REQUESTED** [review]

**Comment by Sairsey** [comment]
Please send me APK of this build

File: src/MangaJet/app/src/main/res/values/strings.xml (hunk 2 above)
Comment (AS2) [reply src/MangaJet/app/src/main/res/values/strings.xml:25]: Ca

File: src/MangaJet/app/src/main/java/com/mangajet/mangajet/mangareader/MangaReaderViewModel.kt [hunk 3]
```diff
@@ -1,31 +1,107 @@
...
+    fun isOnSideChapter() : Boolean {
+        return isOnFirstChapter() || isOnLastChapter()
+    }
 
-    // Function will save current manga state to file
-    private fun saveMangaState() {
-        manga.saveToFile()
+    // Function which check if we are in middle chapter
+    fun isOnMiddleChapter() : Boolean {
+        return !(manga.lastViewedChapter == manga.chapters.size - 1
```
Comment (Sairsey) [inline src/MangaJet/app/src/main/java/com/mangajet/mangajet/mangareader/MangaReaderViewModel.kt:49]: use isOnFirstChapter and isOnLastChapter

File: src/MangaJet/app/src/main/java/com/mangajet/mangajet/mangareader/MangaReaderVPAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+import androidx.recyclerview.widget.RecyclerView
+import com.davemorrissey.labs.subscaleview.ImageSource
+import com.davemorrissey.labs.subscaleview.SubsamplingScaleImageView
+import com.mangajet.mangajet.R
+import com.mangajet.mangajet.data.MangaPage
+
+class MangaReaderVPAdapter(viewModel: MangaReaderViewModel) :
+    RecyclerView.Adapter<MangaReaderVPAdapter.MangaReaderPageHolder>() {
+    companion object {
+        const val SINGLE_CHAPTER_REVERSED_LIST = 0
```
Comment (Sairsey) [inline src/MangaJet/app/src/main/java/com/mangajet/mangajet/mangareader/MangaReaderVPAdapter.kt:16]: Comments

File: src/MangaJet/app/src/main/java/com/mangajet/mangajet/mangareader/MangaReaderVPAdapter.kt [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    inner class MangaReaderPageHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
+        private val imagePage = itemView.findViewById<SubsamplingScaleImageView>(R.id.mangaPage)
+        private val pageAndChapterNumberView =
+            itemView.findViewById<TextView>(R.id.pageAndChapterNumber)
+
+        fun bind(mangaPage : MangaPage, chapter : Int, page : Int) {
+            val chapterOutput = chapter + 1
+            val pageOutput = page + 1
+
+            //pageAndChapterNumberView.text = "Chap: $chapterOutput / Page: $pageOutput"
```
Comment (Sairsey) [inline src/MangaJet/app/src/main/java/com/mangajet/mangajet/mangareader/MangaReaderVPAdapter.kt:33]: it better be in format
"Ch." + chapterOutput ". p." + pageOutput + "/" + totalPages

File: src/MangaJet/app/src/main/java/com/mangajet/mangajet/mangareader/MangaReaderVPAdapter.kt [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+    // viewModel, which contains our interesting data
+    var currentViewModelWithData = viewModel
+
+    inner class MangaReaderPageHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
+        private val imagePage = itemView.findViewById<SubsamplingScaleImageView>(R.id.mangaPage)
+        private val pageAndChapterNumberView =
+            itemView.findViewById<TextView>(R.id.pageAndChapterNumber)
+
+        fun bind(mangaPage : MangaPage, chapter : Int, page : Int) {
```
Comment (Sairsey) [inline src/MangaJet/app/src/main/java/com/mangajet/mangajet/mangareader/MangaReaderVPAdapter.kt:29]: Comments

File: src/MangaJet/app/src/main/java/com/mangajet/mangajet/mangareader/MangaReaderVPAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    companion object {
+        const val SINGLE_CHAPTER_REVERSED_LIST = 0
+        const val SIDE_CHAPTER_REVERSED_LIST   = 1
+        const val MIDDLE_CHAPTER_REVERSED_LIST = 2
+    }
+
+    // viewModel, which contains our interesting data
+    var currentViewModelWithData = viewModel
+
+    inner class MangaReaderPageHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
```
Comment (Sairsey) [inline src/MangaJet/app/src/main/java/com/mangajet/mangajet/mangareader/MangaReaderVPAdapter.kt:24]: Comments

File: src/MangaJet/app/src/main/java/com/mangajet/mangajet/mangareader/MangaReaderVPAdapter.kt [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+                chapterIndex++
+                currentViewModelWithData.manga.chapters[chapterIndex].updateInfo()
+            }
+            else
+                pageIndex = position - 1
+        }
+
+        // leave if bad page index
+        if (pageIndex == -1)
+            return
```
Comment (Sairsey) [inline src/MangaJet/app/src/main/java/com/mangajet/mangajet/mangareader/MangaReaderVPAdapter.kt:99]: do we have such behaviour?

File: src/MangaJet/app/src/main/java/com/mangajet/mangajet/mangareader/MangaReaderViewModel.kt [hunk 9]
```diff
@@ -1,31 +1,107 @@
...
+    }
+
+    // Function which will upload pages
+    fun uploadPages() {
+        for (i in 0 until pagesCount)
+            manga.chapters[manga.lastViewedChapter].getPage(i).upload()
+    }
+
+    // Function which will brutforced reload page
+    fun uploadCurrentPage(page : MangaPage) {
```
Comment (Sairsey) [inline src/MangaJet/app/src/main/java/com/mangajet/mangajet/mangareader/MangaReaderViewModel.kt:79]: better to name reuploadCurrentPage

**Review by Sairsey — CHANGES_REQUESTED** [review]

File: src/MangaJet/app/src/main/java/com/mangajet/mangajet/mangareader/MangaReaderViewModel.kt (hunk 9 above)
Comment (Sairsey) [reply src/MangaJet/app/src/main/java/com/mangajet/mangajet/mangareader/MangaReaderViewModel.kt:79]: And use it

**Review by Sairsey — APPROVED** [review]

**Pull request merged**